### ISSUE TITLE
Document _eval_mpmath

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#23995]
I missed _eval_mpmath when writing [the custom functions guide](https://docs.sympy.org/latest/guides/custom-functions.html#numerical-evaluation-with-evalf). See https://github.com/sympy/sympy/blob/c07abcc94e5b1521062c84d6d0a21b96dedfe07a/sympy/core/function.py#L545-L550

Also, IMO, we ought to use this method everywhere instead of the `MPMATH_TRANSLATIONS` dictionary, which is confusingly repurposed from `lambdify`.

**Comment by oscarbenjamin** [comment]
This function was added in https://github.com/sympy/sympy/pull/20569. It's original purpose was to handle the situation where the `args` of a function don't map directly to the args of any particular mpmath function. It could be used in general though to specify how to evalf a function since it doesn't suffer some of the deficiencies of `_eval_evalf` (e.g. #20479, #20493).

**Comment by asmeurer** [comment]
And also there's a lot of logic in `Function._eval_evalf` that shouldn't be duplicated in the subclasses (see https://github.com/sympy/sympy/pull/23992#discussion_r958926963).

**Comment by asmeurer** [comment]
One question is which should be preferred, recursively evalfing another expression, or returning a mpmath function? For example, for the `versin` example in the guide, it currently says to do

```py
from sympy import sin
class versin(Function):
    def _eval_evalf(self, prec):
        return (2*sin(self.args[0]/2)**2)._eval_evalf(prec)
```

But it could also do

```py
import mpmath
class versin(Function):
    def _eval_mpmath(self):
        return lambda x: 2*mpmath.sin(self.args[0]/2)**2, self.args
```

It would be good to note the pros and cons of each approach in the guide.

**Comment by oscarbenjamin** [comment]
Unfortunately `_eval_mpmath` is only called from `Function._eval_evalf`. The `_eval_evalf` signature is broken so it cannot be used correctly in general.

What is really needed is two separate methods:

1. A method that handles manipulating precision and all the other options for evalf including controlling the precision of recursive evaluation. This is the method to call when evaluating recursively.
2. A method that simply specifies the appropriate mpmath routine (and args) but doesn't explicitly handle precision (because it should be called by something that does).

Then core things like Add, Pow etc can use 1. but most functions would use 2. and `Function` would define 1. to call 2.

The `_eval_mpmath` method performs 2. but there is no proper method for 1. since that is currently handled by the `evalf_*` functions in `sympy/core/evalf.py` (and not really handled properly for anything else).

**Comment by asmeurer** [comment]
I agree that it would make sense to move it from `Function` to `Expr`. There is a lot of code in Function that could be moved to `Expr`, or even `Basic`. The entire `eval` logic instead of explicitly defining `__new__`, for instance. As far as the custom functions guide goes it doesn't matter because it specifically focuses on `Function` subclasses, so it's already relevant with the current code to consider how to document this there.

**Comment by asmeurer** [comment]
Here's a good example for the issues you describe in #20479. Suppose we weren't careful in the `versin` example and defined evalf using `1 - cos(x)`, which suffers from loss of significance when cos(x) is near 1. 

```py
>>> class versin(Function):
...     def _eval_evalf(self, prec):
...         return (1 - cos(self.args[0]))._eval_evalf(prec)
>>> versin(S(1)/2**100).evalf()
0
>>> versin(S(1)/2**100).evalf(60)
3.11150763893057085357203202689006212029512608436058356655056e-61
```

The same problem occurs with `_eval_mpmath`.

```
>>> class versin(Function):
...     def _eval_mpmath(self):
...         return lambda x: 1 - mpmath.cos(x), self.args
>>> versin(S(1)/2**100).evalf()
0
>>> versin(S(1)/2**100).evalf(60)
3.11150763893057085357203202689006212029512608436058356655056e-61
```

However, `evalf` itself gets it right

```
>>> (1 - cos(S(1)/2**100)).evalf()
3.11150763893057e-61
>>> (1 - cos(x)).evalf(subs={x: S(1)/2**100})
3.11150763893057e-61
```

So until `_eval_evalf` gets fixed, there isn't actually a difference between [the above two ways of implementing evalf](https://github.com/sympy/sympy/issues/23995#issuecomment-1232211699). But once it is then `_eval_evalf` will be better in this case, because returning a  `lambda` from `_eval_mpmath` makes it skip the `evalf_add` logic that gives the right answer here. And of course, you really should avoid the numerically unstable expression `1 - cos(x)` entirely in favor of `2*sin(x/2)**2` (but that's not always easy for every function).

So once these issues are fixed and `_eval_evalf` were made to use all the machinery of `evalf`, then the advice we should give in the guide would be:

- If the function has the same name as an mpmath function you don't need to do anything.
- If the function is in mpmath but has a different name, define `_eval_mpmath`.
- If you can recursively evalf a different expression, implement `_eval_evalf` (or whatever method we end up with after fixing https://github.com/sympy/sympy/issues/20479) recursively. Do *not* implement `_eval_mpmath` to return anything but a single function, as that will bypass the evalf code that handles things like Add intelligently. Also, try to use an expression that is numerically stable. 
- In general, you may need to call the helper functions in evalf.py, and functions in mpmath. But be careful about calling mpmath directly because it will bypass the helper functions in evalf.py which automatically increase precision to avoid loss of significance. 

For `versin`, that means recursively calling evalf, and using the `2*sin(x/2)**2` formula. Basically, `_eval_mpmath` should only ever be used when the function is in mpmath, but under a different name (or with a different name for different number of arguments, like with `euler` in https://github.com/sympy/sympy/pull/23992#discussion_r958926963). 

That's once https://github.com/sympy/sympy/issues/20479 is fixed. For now, I'm not sure what to recommend. I guess what is written there already is fine.
